### ISSUE TITLE
Windows file changes are not captured when dockerized

### ISSUE THREAD

**Issue by hihaha99** [cosmtrek/air#190]
The changes made on windows are reflected inside the docker container via volumes but not capture by Air. Editing the file directly inside the docker will trigger rebuild, changes reflected on windows side also.

**Comment by xEtarusx** [comment]
Got the same issue. When editing files with Goland on Windows it does not trigger the rebuild.
Any ideas on how to solve this problem?

**Comment by makiuchi-d** [comment]
If you are running the docker engine on the WSL2, this behavior is due to the WSL2 limitation.
inotify filesystem events are not propagated between WSL2 and Windows.
https://github.com/microsoft/WSL/issues/4739

You can avoid inotify problem by environment variable `CHOKIDAR_USEPOLLING=true`, but it may cause performance problem.

**Comment by hihaha99** [comment]
> If you are running the docker engine on the WSL2, this behavior is due to the WSL2 limitation.
> inotify filesystem events are not propagated between WSL2 and Windows.
> [microsoft/WSL#4739](https://github.com/microsoft/WSL/issues/4739)

Thanks, that's the source I think.

> You can avoid inotify problem by environment variable `CHOKIDAR_USEPOLLING=true`, but it may cause performance problem.

CHOKIDAR is for nodejs from what I searched. Any solution for go?

**Comment by makiuchi-d** [comment]
ah... sorry, air uses fsnotify, not chokidar.
as far as i know, air (and fsnotify) does not implement polling fallback.
so it seems that this problem cannot be avoided.

**Comment by CRYBOII** [comment]
Can anyone  solve this problem ?

**Comment by NikosDevPhp** [comment]
We are still waiting for WSL solution. Until then:

Put your Project from Windows to `\\wsl$` -> {Your Distro} -> /home -> {username} / {[project_path} and NOT in /mnt/c/Users/...

Then Open the first location (\\wsl$ from VSCode or editor of your choice). Save will now send the notification of the changed file

NTFS vs linux fs issue

**Comment by CRYBOII** [comment]
> We are still waiting for WSL solution. Until then:
> 
> Put your Project from Windows to `\\wsl$` -> {Your Distro} -> /home -> {username} / {[project_path} and NOT in /mnt/c/Users/...
> 
> Then Open the first location (\wsl$ from VSCode or editor of your choice). Save will now send the notification of the changed file
> 
> NTFS vs linux fs issue

I can't find \\wsl$

**Comment by CoolersCoder** [comment]
@CRYBOII  works for me
The idea is enable WSL2 in your windows system, and install any Linux system such as Ubuntu.
1. wsl --set-version Ubuntu 2
2. Refresh docker setting -> Resouces -> WSL INTEGRATION -> enable Ubuntu
3. open terminal and type wsl
4. cp -r /mnt/c/Users/XXXX/projectname  ~/
5. code .
6. docker-compose up --build

**Comment by CoolersCoder** [comment]
> We are still waiting for WSL solution. Until then:
> 
> Put your Project from Windows to `\\wsl$` -> {Your Distro} -> /home -> {username} / {[project_path} and NOT in /mnt/c/Users/...
> 
> Then Open the first location (\wsl$ from VSCode or editor of your choice). Save will now send the notification of the changed file
> 
> NTFS vs linux fs issue

wsl --set-version Ubuntu 2
Refresh docker setting -> Resouces -> WSL INTEGRATION -> enable Ubuntu
open terminal and type wsl
cp -r /mnt/c/Users/XXXX/projectname ~/
code .
docker-compose up --build


> We are still waiting for WSL solution. Until then:
> 
> Put your Project from Windows to `\\wsl$` -> {Your Distro} -> /home -> {username} / {[project_path} and NOT in /mnt/c/Users/...
> 
> Then Open the first location (\wsl$ from VSCode or editor of your choice). Save will now send the notification of the changed file
> 
> NTFS vs linux fs issue

Thanks for sharing info, I get more details to resolve windows system docker issue.
The idea is enable WSL2 in your windows system, and install any Linux system such as Ubuntu.

wsl --set-version Ubuntu 2
Refresh docker setting -> Resouces -> WSL INTEGRATION -> enable Ubuntu
open terminal and type wsl
cp -r /mnt/c/Users/XXXX/projectname ~/
code .
docker-compose up --build

**Comment by hihaha99** [comment]
> > We are still waiting for WSL solution. Until then:
> > Put your Project from Windows to `\\wsl$` -> {Your Distro} -> /home -> {username} / {[project_path} and NOT in /mnt/c/Users/...
> > Then Open the first location (\wsl$ from VSCode or editor of your choice). Save will now send the notification of the changed file
> > NTFS vs linux fs issue
> 
> Thanks for sharing info, I get more details to resolve windows system docker issue. The idea is enable WSL2 in your windows system, and install any Linux system such as Ubuntu.
> 
> 1. wsl --set-version Ubuntu 2
> 2. Refresh docker setting -> Resouces -> WSL INTEGRATION -> enable Ubuntu
> 3. open terminal and type wsl
> 4. cp -r /mnt/c/Users/XXXX/projectname ~/
> 5. code .
> 6. docker-compose up --build

Can confirm this works. Thank you!

**Comment by yigeqiji** [comment]
Is this problem solved?

**Comment by insyri** [comment]
> Is this problem solved?

No, right now the solution is to move the project into the Linux filesystem in WSL.

See https://github.com/cosmtrek/air/issues/190#issuecomment-999254716

**Comment by yigeqiji** [comment]
> 
thanks

**Comment by yigeqiji** [comment]
> > Is this problem solved?
> 
> No, right now the solution is to move the project into the Linux filesystem in WSL.
> 
> See [#190 (comment)](https://github.com/cosmtrek/air/issues/190#issuecomment-999254716)

thanks，i will take a look，hope this problem can be solved as soon as possible

**Comment by vkhobor** [comment]
Hi. I got the same issue, too.

Hugo has a fallback polling mechanism.
https://github.com/gohugoio/hugo/pull/8723
Could this be implemented here too ?

**Comment by insyri** [comment]
> Hi. I got the same issue, too.
> 
> Hugo has a fallback polling mechanism implemented. [gohugoio/hugo#8723](https://github.com/gohugoio/hugo/pull/8723) Could something like this be done here too ?

Air uses [fsnotify](https://github.com/fsnotify/fsnotify), which is a go library for listening to file system events. fsnotify has not implemented a polling fallback feature. The issue (https://github.com/fsnotify/fsnotify/issues/9) has been in discussion since 2014; as of right now, I would recommend using an alternative like stated above until further notice.

**Comment by xiantang** [comment]
Thanks for reporting this. Quick update for anyone finding this issue today:

 This behavior is mostly from WSL2/Windows filesystem event limits (inotify events from Windows-mounted
 paths are not always propagated into Linux containers), so Air may not see file changes when the
 project is on /mnt/c/....

 Air now supports polling mode as a workaround. In your .air.toml:

```
 [build]
   poll = true
   poll_interval = 500
```

 You can also reduce this problem by keeping the project in the Linux filesystem (for example under \\
 wsl$\<distro>\home\<user>\...) instead of Windows-mounted paths.

 If people still see this issue with poll = true, please share:
 - your .air.toml
 - project path (/mnt/c/... vs Linux path)
 - Docker/WSL2 versions
 - a small repro repo
